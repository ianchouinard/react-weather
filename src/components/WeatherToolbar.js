import React, {Component} from 'react'
import background from './../img/header_bg.jpg'

export class WeatherToolbar extends Component {

  state = {
    zipCode: '03063',
    zipLooksValid: true
  }

  /**
   * Update the zipcode state.
   */
  zipChanged = (e) => this.setState({zipCode: e.target.value})

  /**
   * Emit the entered zip code.
   */
  zipEntered = (e) => {
    e.preventDefault()
    this.props.zipEntered(this.state.zipCode)
  }

  /**
   * Attempt to determine if the zip looks valid.
   * Will be disabled if true
   */
  zipLooksInvalid = () => {
    if (isNaN(parseInt(this.state.zipCode))) {
      return true
    }

    if (this.state.zipCode.length !== 5) {
      return true
    }

    return false
  }

  render() {
    return(
      <section className="weatherToolbar" style={{backgroundImage: `url(${background})`}}>
        <div className="container">
          <form className="toolbar">
            <h1>Find Your Local Weather</h1>
            
            <div className="form-group">
              <label>Zip Code</label>
              <input
                type="number"
                placeholder="zip code"
                value={this.state.zipCode}
                onChange={this.zipChanged} />
              <button
                type="submit"
                disabled={this.zipLooksInvalid()}
                onClick={this.zipEntered}>Get Weather</button>
            </div>
          </form>
        </div>
      </section>
    )
  }

}
