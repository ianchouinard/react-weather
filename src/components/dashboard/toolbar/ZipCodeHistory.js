import React, { Component } from 'react'
import { MenuToggle } from '../../MenuToggle'
import { store } from '../../../store/store';
import { setZipCode } from '../../../actions/ForecastActions';

export class ZipCodeHistory extends Component {

  state = {
    menuOpen: false,
    zipCodeHistory: store.getState().zipCodeHistory
  }

  componentDidMount() {
    store.subscribe(() => {
      this.setState({
        zipCodeHistory: store.getState().zipCodeHistory
      })
    })
  }

  /**
   * User selected a zip code. 
   * Emit the zip code to the parent component
   */
  zipSelected = (e, zip) => {
    e.preventDefault()
    store.dispatch(setZipCode(zip))
  }

  /**
   * Toggle the mobile menu state
   */
  toggleMenu = () => {
    const menuState = this.state.menuOpen

    this.setState({
      menuOpen: !menuState
    })
  }

  render() {
    return (
      <nav className="zipCodeHistory">
        <MenuToggle open={this.state.menuOpen} toggleClick={this.toggleMenu} />
        <span>Recent Locations</span>
        <ul className={(this.state.menuOpen ? 'mobileOpen' : '')}>
          {this.state.zipCodeHistory.map((zipCode, index) => {
            return (
              <li key={ index }>
                <a
                  href="/"
                  title={`View weather for ${zipCode}`}
                  onClick={(e) => {this.zipSelected(e, zipCode)}}>{ zipCode }</a>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }

}
