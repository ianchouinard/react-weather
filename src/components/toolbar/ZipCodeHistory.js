import React, { Component } from 'react'
import { MenuToggle } from '../MenuToggle'

export class ZipCodeHistory extends Component {

  state = {
    menuOpen: false
  }

  zipSelected = (e, zip) => {
    e.preventDefault()
    this.props.zipEntered(zip)
  }

  toggleMenu = () => {
    const menuState = this.state.menuOpen

    this.setState({
      menuOpen: !menuState
    })
  }

  render() {
    return (
      <nav className="zipCodeHistory">
        <MenuToggle open={this.state.menuOpen} toggleClick={this.toggleMenu} />
        <span>Recent Locations</span>
        <ul className={(this.state.menuOpen ? 'mobileOpen' : '')}>
          {this.props.zipCodeHistory.map((zipCode, index) => {
            return (
              <li key={ index }>
                <a
                  href="/"
                  title={`View weather for ${zipCode}`}
                  onClick={(e) => {this.zipSelected(e, zipCode)}}>{ zipCode }</a>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }

}
