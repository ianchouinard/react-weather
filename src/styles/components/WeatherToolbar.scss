
.weatherToolbar {
  position: relative;
  padding-top: 40px;
  background-size: cover;
  background-position: center;
  z-index: 10;

  .toolbar {
    padding: 25px 10px 10px 10px;
    text-align: center;
    background: #F7F7F7;
    box-shadow: -2px 3px 3px -1px rgba(0, 0, 0, 0.15);
    transform: translateY(50%);

    h1 {
      margin-bottom: 20px;
      font-size: $font_large;
    }

    .form-group {
      label {
        margin-right: 10px;
      }

      input {
        padding: 5px;
        margin-right: 10px;
        border: none;
        box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.25)
      }
    }
  }

  @media screen and (max-width: 500px) {
    padding-top: 0;
    margin-bottom: 0;
    background-image: none;

    .container {
      padding: 0;
    }

    .toolbar {
      transform: translateY(0);

      .form-group {
        input {
          display: block;
          max-width: none;
          width: 100%;
          margin-bottom: 10px;
        }
      }
    }
  }
}
