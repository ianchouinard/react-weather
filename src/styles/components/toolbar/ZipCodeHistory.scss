
.zipCodeHistory {
  position: relative;
  display: flex;
  flex-wrap: nowrap;
  padding-bottom: 5px;
  height: 38px;
  overflow-y: hidden;
  z-index: 2;
  color: #fff;
  border-bottom: 1px solid #fff;

  .menuToggle {
    display: none;
    margin-right: 10px;

    &+span {
      vertical-align: middle;
    }
  }

  span {
    margin-right: 10px;
    font-weight: 700;
    font-size: 1.2rem;
    white-space: nowrap;
  }

  ul {
    display: flex;
    flex-wrap: nowrap;
    vertical-align: middle;
    margin: 0;
    padding-left: 10px;
    min-height: 58px;
    overflow-x: auto;
    list-style: none;

    li {
      margin-right: 7px;

      a {
        display: inline-block;
        padding: 5px 10px;
        font-size: .9rem;
        color: #fff;
        border-radius: 3px;

        &:hover {
          background: rgba(250,250,250,.15);
          text-decoration: none;
        }
      }
    }
  }

  @media screen and (max-width: $mobile_small) {
    display: block;
    padding: 15px 0 5px 0;
    height: auto;
    border-bottom: none;
    background: $color_grey;

    .menuToggle {
      display: inline-block;
      margin-left: 15px;
    }

    ul {
      max-height: 0;
      min-height: 0;
      overflow: hidden;
      padding: 0 15px;
      margin-top: 10px;
      display: block;
      transition: all 350ms;

      &.mobileOpen {
        max-height: 200px;
        overflow-y: auto;
      }

      li {
        display: block;
        width: 100%;
        margin: 0;
      }
    }
  }
}
