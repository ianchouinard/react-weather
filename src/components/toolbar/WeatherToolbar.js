import React, {Component} from 'react'
import { ZipCodeHistory } from './ZipCodeHistory'
import background from './../../img/header_bg.jpg'
import backgroundRain from './../../img/header_bg_rain.jpg'
import backgroundSnow from './../../img/header_bg_snow.jpg'

export class WeatherToolbar extends Component {

  state = {
    zipCode: '03063',
    zipLooksValid: true
  }

  componentWillReceiveProps(props) {
    this.setState({
      zipCode: props.zipCode
    })
  }

  /**
   * Update the zipcode state.
   */
  zipChanged = (e) => this.setState({zipCode: e.target.value})

  /**
   * Emit the entered zip code.
   */
  zipEntered = (e) => {
    e.preventDefault()
    this.props.zipEntered(this.state.zipCode)
  }

  /**
   * Attempt to determine if the zip looks valid.
   * Will be disabled if true
   */
  zipLooksInvalid = () => {
    if (isNaN(parseInt(this.state.zipCode))) {
      return true
    }

    if (this.state.zipCode.length !== 5) {
      return true
    }

    return false
  }

  /**
   * Get header background based on weather condtions
   * weather code info here 
   * https://openweathermap.org/weather-conditions
   */
  getImageSource = () => {
    if (!this.props.conditions || !this.props.conditions.weather) {
      return ''
    }
    
    const code = this.props.conditions.weather[0].id
    if (code >= 700) {
      return background // normal
    } else if (code >= 600) {
      return backgroundSnow // snow
    } else if (code >= 200) {
      return backgroundRain // rain
    } else {
      return background
    }
  }

  render() {
    return(
      <section className="weatherToolbar" style={{backgroundImage: `url(${this.getImageSource()})`}}>
        <div className="container">
          <ZipCodeHistory
            zipCodeHistory={this.props.zipCodeHistory}
            zipEntered={this.props.zipEntered} />

          <form className="toolbar">
            <h1>Find Your Local Weather</h1>
            
            <div className="form-group">
              <label>Zip Code</label>
              <input
                type="text"
                placeholder="zip code"
                value={this.state.zipCode}
                onChange={this.zipChanged} />
              <button
                type="submit"
                disabled={this.zipLooksInvalid()}
                onClick={this.zipEntered}
                title="Get forecast for this location">Get Weather</button>
            </div>
          </form>
        </div>
      </section>
    )
  }

}
