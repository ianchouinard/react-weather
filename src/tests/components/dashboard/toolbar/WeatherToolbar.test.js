import React from 'react';
import renderer from 'react-test-renderer';
import { WeatherToolbar } from './../../../../components/dashboard/toolbar/WeatherToolbar'

describe('WeatherToolbar', () => {
  const component = renderer.create(
    <WeatherToolbar
      zipCode={'03063'}
      conditions={{}} />
  )

  it('should create the component', () => {
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('zipLooksInvalid()', () => {
    it('should return false if zipcode looks valid', () => {
      const zipTestValid = component.getInstance().zipLooksInvalid()
      expect(zipTestValid).toBe(false)
    })

    it('should return true if zipcode looks invalid', () => {
      component.getInstance().setState({zipCode: 'nope'})
      const zipTestInvalid = component.getInstance().zipLooksInvalid()
      expect(zipTestInvalid).toBe(true)
      component.getInstance().setState({zipCode: '03063'})
    })
  })

  describe('getImageSource()', () => {

    const mockCondition = (id) => {
      return {
        weather: [
          {
            id: id
          }
        ]
      }
    }

    it('should figure out an appropriate image source', () => {
      const component1 = renderer.create(
        <WeatherToolbar
          zipCode={'03063'}
          conditions={mockCondition(800)} />
      )
      const test1 = component1.getInstance().getImageSource()
      expect(test1).toBe('header_bg.jpg')

      const component2 = renderer.create(
        <WeatherToolbar
          zipCode={'03063'}
          conditions={mockCondition(600)} />
      )
      const test2 = component2.getInstance().getImageSource()
      expect(test2).toBe('header_bg_snow.jpg')

      const component3 = renderer.create(
        <WeatherToolbar
          zipCode={'03063'}
          conditions={mockCondition(200)} />
      )
      const test3 = component3.getInstance().getImageSource()
      expect(test3).toBe('header_bg_rain.jpg')
    })
  })
})
