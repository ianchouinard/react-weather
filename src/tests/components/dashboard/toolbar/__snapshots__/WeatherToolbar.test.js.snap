// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherToolbar should create the component 1`] = `
<section
  className="weatherToolbar"
  style={
    Object {
      "backgroundImage": "url()",
    }
  }
>
  <div
    className="container"
  >
    <nav
      className="zipCodeHistory"
    >
      <button
        className="menuToggle "
        onClick={[Function]}
        title="Open Menu"
      >
        <span />
        <span />
        <span />
      </button>
      <span>
        Recent Locations
      </span>
      <ul
        className=""
      />
    </nav>
    <form
      className="toolbar"
    >
      <h1>
        Find Your Local Weather
      </h1>
      <div
        className="form-group"
      >
        <label>
          Zip Code
        </label>
        <input
          onChange={[Function]}
          placeholder="zip code"
          type="text"
          value="03063"
        />
        <button
          disabled={false}
          onClick={[Function]}
          title="Get forecast for this location"
          type="submit"
        >
          Get Weather
        </button>
      </div>
    </form>
  </div>
</section>
`;
